File: rust/src/code/insert_interval.rs
pub struct Solution {}

impl Solution {
    pub fn insert(intervals: Vec<Vec<i32>>, mut new_interval: Vec<i32>) -> Vec<Vec<i32>> {
        let mut res: Vec<Vec<i32>> = Vec::new();
        for i in 0..intervals.len() {
            if new_interval[1] < intervals[i][0] {
                res.push(new_interval.clone());
                res.append(&mut intervals[i..].to_vec());
                return res;
            } else if new_interval[0] > intervals[i][1] {
                res.push(intervals[i].clone());
            } else {
                new_interval = Vec::from([
                    std::cmp::min(new_interval[0], intervals[i][0]),
                    std::cmp::max(new_interval[1], intervals[i][1]),
                ]);
            }
        }

        res.push(new_interval);
        return res;
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn case_1() {
        let intervals = vec![vec![1, 3], vec![6, 9]];
        let new_interval = vec![2, 5];
        let expected = vec![vec![1, 5], vec![6, 9]];
        let actual = Solution::insert(intervals, new_interval);

        assert_eq!(expected, actual);
    }

    #[test]
    fn case_2() {
        let intervals = vec![
            vec![1, 2],
            vec![3, 5],
            vec![6, 7],
            vec![8, 10],
            vec![12, 16],
        ];
        let new_interval = vec![4, 8];
        let expected = vec![vec![1, 2], vec![3, 10], vec![12, 16]];
        let actual = Solution::insert(intervals, new_interval);

        assert_eq!(expected, actual);
    }
}

File: python/insert_interval.py
class Solution:
    def insert(
        self, intervals: list[list[int]], newInterval: list[int]
    ) -> list[list[int]]:
        """tiem complexity: O(n)"""
        result: list[list[int]] = []
        for i in range(len(intervals)):
            if intervals[i][0] > newInterval[1]:
                result.append(newInterval)
                return result + intervals[i:]
            elif intervals[i][1] < newInterval[0]:
                result.append(intervals[i])
            else:
                newInterval = [
                    min(intervals[i][0], newInterval[0]),
                    max(intervals[i][1], newInterval[1]),
                ]
        result.append(newInterval)
        return result


def test_insert_case_1():
    # arrange
    intervals: list[list[int]] = [[1, 3], [6, 9]]
    newInterval: list[list[int]] = [2, 5]
    expected: list[list[int]] = [[1, 5], [6, 9]]

    # act
    solution = Solution()
    actual = solution.insert(intervals, newInterval)

    # assert
    assert expected == actual


def test_insert_case_2():
    # arrange
    intervals: list[list[int]] = [[1, 2], [3, 5], [6, 7], [8, 10], [12, 16]]
    newInterval: list[list[int]] = [4, 8]
    expected: list[list[int]] = [[1, 2], [3, 10], [12, 16]]

    # act
    solution = Solution()
    actual = solution.insert(intervals, newInterval)

    # assert
    assert expected == actual


def test_insert_case_3():
    # arrange
    intervals: list[list[int]] = []
    newInterval: list[list[int]] = [5, 7]
    expected: list[list[int]] = [[5, 7]]

    # act
    solution = Solution()
    actual = solution.insert(intervals, newInterval)

    # assert
    assert expected == actual
